src/types.rs

```diff
@@ -101,16 +138,23 @@ impl TypeValidator for TypeDefinition {
         match self {
             TypeDefinition::Named(named_type) => named_type.validate(value, type_store),
             TypeDefinition::Anonymous(anonymous_type) => anonymous_type.validate(value, type_store),
-            TypeDefinition::Core(ion_type) => {
-                if value.ion_type() != *ion_type {
-                    return Err(Violation::new(
-                        "type_constraint",
-                        ViolationCode::TypeMismatched,
-                        &format!("expected type {:?}, found {:?}", ion_type, value.ion_type()),
-                    ));
+            TypeDefinition::Core(core_type) => match core_type {
+                CoreISLTypeDefinition::IonType(ion_type) => {
+                    if value.ion_type() != *ion_type {
+                        return Err(Violation::new(
+                            "type_constraint",
+                            ViolationCode::TypeMismatched,
+                            &format!("expected type {:?}, found {:?}", ion_type, value.ion_type()),
+                        ));
+                    }
+                    return Ok(());
                 }
-                return Ok(());
-            }
+                CoreISLTypeDefinition::Other(other_type) => {
+                    let type_id = type_store.get_core_type_id(core_type).unwrap();
+                    let type_def = type_store.get_type_by_id(*type_id).unwrap();
+                    type_def.validate(value, type_store)
+                }
```

Comment:
This is basically the same logic as what we have for `TypeDefinition::Named::validate(value, typestore)`. I think we're going to run into maintenance problems if we have two separate code paths for the `Other` and `Named` types. Can we (internally) represent types like `any` and `lob` as `Named` types instead of having `Other`?

Reply:
Only reason to separate out `Others` from `Named` types is to have `core_types` stored in a separate `hashmap` inside `TypeStore`/ `PendingTypes`. This will eventually be helpful for getting types defined within a `schema` to not include the `core_types`.

Reply:
If we're storing them in separate places, we could still use the same `TypeDefinition` implementation, right? We'll still have separate code for retrieving the types, but we don't need to have duplicate code for evaluating them.

Reply:
Yes. `TypeDefinition` is still the same but this is an enum variant of `TypeDefinition`. 
```
pub enum TypeDefinition {
    Named(TypeDefinitionImpl),
    Anonymous(TypeDefinitionImpl),
    Core(CoreISLTypeDefinition),
}

```
But I do understand this should not be exposed to public APIs and hence I can change the methods of `TypeStore` to use `TypeDefinition` for adding and getting types instead of `CoreISLTypeDefinition`. Does this sound about right?

Reply:
Preloading all the builtin types to `TypeStore`.